src/main/java/net/dries007/tfc/common/blocks/devices/BloomeryBlock.java

```diff
@@ -0,0 +1,243 @@
+            .matchEachDirection(origin.east(), stoneMatcher, new Direction[]{Direction.NORTH, Direction.SOUTH, Direction.DOWN}, 1)
+            .matchHorizontal(origin.above(), stoneMatcher, 1);
+
+        BLOOMERY_CHIMNEY = new MultiBlock()
+            .match(origin, insideChimney)
+            .matchHorizontal(origin, stoneMatcher, 1);
+
+        // Gate center is the bloomery gate block
+        GATE_Z = new MultiBlock()
+            .match(origin, state -> state.is(TFCBlocks.BLOOMERY.get()) || state.isAir())
+            .matchEachDirection(origin, stoneMatcher, new Direction[]{Direction.WEST, Direction.EAST, Direction.UP, Direction.DOWN}, 1);
+
+        GATE_X = new MultiBlock()
+            .match(origin, state -> state.is(TFCBlocks.BLOOMERY.get()) || state.isAir())
+            .matchEachDirection(origin, stoneMatcher, new Direction[]{Direction.NORTH, Direction.SOUTH, Direction.UP, Direction.DOWN}, 1);
+    }
+
+    public static int getChimneyLevels(Level level, BlockPos centerPos)
+    {
+        for (int i = 1; i < 4; i++)
+        {
+            BlockPos center = centerPos.above(i);
+            if (!BLOOMERY_CHIMNEY.test(level, center))
+            {
+                return i - 1;
+            }
+        }
+        // Maximum levels
+        return 3;
+    }
+
+    public BloomeryBlock(ExtendedProperties properties)
+    {
+        super(properties);
+        /* todo: deal with these from 1.12
+         * setHarvestLevel("pickaxe", 0); -> requiresCorrectToolForDrops + json?
+         * setHardness(20.0F);
+         */
+        registerDefaultState(getStateDefinition().any()
+            .setValue(FACING, Direction.NORTH)
+            .setValue(LIT, false)
+            .setValue(OPEN, false));
+    }
+
+    public boolean canGateStayInPlace(Level level, BlockPos pos, Direction.Axis axis)
+    {
+        if (axis == Direction.Axis.X)
+        {
+            return GATE_X.test(level, pos);
+        }
+        else
+        {
+            return GATE_Z.test(level, pos);
+        }
+    }
+
+    //if Directions don't have int indices any more, then...
+    public boolean isFormed(Level level, BlockPos centerPos, Direction facing)
+    {
+        return switch (facing)
+            {
+                case NORTH -> BLOOMERY_BASE[0].test(level, centerPos);
+                case SOUTH -> BLOOMERY_BASE[1].test(level, centerPos);
+                case WEST -> BLOOMERY_BASE[2].test(level, centerPos);
+                case EAST -> BLOOMERY_BASE[3].test(level, centerPos);
+                default -> false;
+            };
+    }
+
+    /* todo: find out if there's a modern version of following methods
+     * getStateFromMeta, getMetaFromState, isFullCube
+     * getBoundingBox, getBlockFaceShape, getCollisionBoundingBox, isOpaqueCube
+     * collisionRayTrace, canPlaceBlockAt (canSurvive?), hasTileEntity
+     * and does TFC still have:
+     * onBreakBlock
+     */
```

Comment:
onBreakBlock -> onRemove

The 'full cube' thing and its friend are gone. Just set your stuff up in getShape and be done with it.
Metadata is gone. Use createBlockStateDefinition and that handles it
canPlaceBlockAt -> if false, return AIR in getStateForPlacement
hasTileEntity -> taken care of in the ExtendedProperties